app/src/main/java/fr/neamar/kiss/cache/MemoryCacheHelper.java

```diff
@@ -0,0 +1,165 @@
+     * @param className  app
+     * @param userHandle android.os.UserHandle
+     * @param callback   this will be called when we know what the drawable is
+     */
+    public static <T> void getAppIconDrawable(@NonNull Context context, ComponentName className, UserHandle userHandle, @Nullable OnDrawableReady<T> callback) {
+        AppIconHandle handle = new AppIconHandle(className, userHandle);
+        Drawable drawable;
+        synchronized (sAppIconCache) {
+            drawable = sAppIconCache.get(handle);
+            if (drawable == null) {
+                // if the drawable cached for this app is null don't bother trying to find it again
+                if (!sAppIconCache.containsKey(handle)) {
+                    new AsyncAppIconLoad(context, handle, callback).execute();
+                    return;
+                }
+            }
+        }
+        if (callback != null)
+            callback.onDrawableReady(drawable);
+    }
+
+    /**
+     * If the app icon is not found in the cache we load it. Else return cache value. Synchronous function.
+     *
+     * @param context    context to use
+     * @param className  app
+     * @param userHandle android.os.UserHandle
+     * @return app icon drawable from cache
+     */
+    public static Drawable getAppIconDrawable(@NonNull Context context, ComponentName className, UserHandle userHandle) {
+        AppIconHandle handle = new AppIconHandle(className, userHandle);
+        return getAppIconDrawable(context, handle);
+    }
+
+    /**
+     * This is called from the async task. To prevent accidents, do not access sAppIconCache without a sync block
+     *
+     * @param context context to use
+     * @param handle  wrapper for app name
+     * @return app icon drawable from cache
+     */
+    private static Drawable getAppIconDrawable(@NonNull Context context, AppIconHandle handle) {
+        Drawable drawable;
+        synchronized (sAppIconCache) {
+            drawable = sAppIconCache.get(handle);
+            if (drawable == null) {
+                // if the drawable cached for this app is null don't bother trying to find it again
+                if (sAppIconCache.containsKey(handle))
+                    return null;
+                drawable = KissApplication.getApplication(context).getIconsHandler()
+                        .getDrawableIconForPackage(handle.componentName, handle.userHandle);
+                sAppIconCache.put(handle, drawable);
+            }
+        }
+        return drawable;
+    }
+
+    public static void cacheAppIconDrawable(@NonNull Context context, ComponentName className, UserHandle userHandle) {
+        AppIconHandle handle = new AppIconHandle(className, userHandle);
+        if (!sAppIconCache.containsKey(handle))
+            new AsyncAppIconLoad(context, handle, null).execute();
+    }
+
+    public static void trimMemory() {
+        synchronized (sAppIconCache) {
+            sAppIconCache.clear();
+        }
+        System.gc();
```

Comment:
[Never](https://stackoverflow.com/a/2414120/3124150).

Reply:
It seemed like a good "free memory" thing to do, but it's not necessary